File: python/examples/webhook_server.py
#!/usr/bin/env python3
"""
Webhook Server Example for JewelMusic Python SDK

This example demonstrates how to set up and handle JewelMusic webhooks:
- Creating webhook endpoints
- Verifying webhook signatures  
- Handling different event types
- Testing webhooks
- Managing webhook deliveries
"""

import os
import asyncio
import json
import secrets
from datetime import datetime
from typing import Dict, List, Optional, Any
from flask import Flask, request, jsonify
import threading
import signal
import sys

from jewelmusic_sdk import JewelMusic
from jewelmusic_sdk.webhooks import WebhookResource
from jewelmusic_sdk.exceptions import (
    AuthenticationError,
    ValidationError,
    NetworkError
)


class WebhookServer:
    """JewelMusic webhook server implementation."""
    
    def __init__(self, api_key: Optional[str] = None, port: int = 8080):
        """Initialize webhook server."""
        self.api_key = api_key or os.getenv('JEWELMUSIC_API_KEY')
        if not self.api_key:
            raise ValueError("API key required. Set JEWELMUSIC_API_KEY environment variable.")
        
        self.client = JewelMusic(
            api_key=self.api_key,
            environment='sandbox'
        )
        
        self.port = port
        self.webhook_path = '/webhooks/jewelmusic'
        self.webhook_secret = self._generate_webhook_secret()
        self.app = Flask(__name__)
        self.webhook_id = None
        
        # Setup Flask routes
        self._setup_routes()
    
    def _generate_webhook_secret(self) -> str:
        """Generate secure webhook secret."""
        return secrets.token_urlsafe(32)
    
    def _setup_routes(self):
        """Setup Flask application routes."""
        
        @self.app.route('/', methods=['GET'])
        def home():
            """Home page with webhook information."""
            return f"""
            <!DOCTYPE html>
            <html>
            <head>
                <title>JewelMusic Webhook Server</title>
                <style>
                    body {{ font-family: Arial, sans-serif; margin: 40px; }}
                    .header {{ color: #333; }}
                    .endpoint {{ background: #f5f5f5; padding: 10px; border-radius: 5px; }}
                    .status {{ color: #28a745; }}
                </style>
            </head>
            <body>
                <h1 class="header">🎵 JewelMusic Webhook Server</h1>
                <p class="status">✅ Server is running and ready to receive webhooks</p>
                
                <h2>Endpoints:</h2>
                <div class="endpoint">
                    <strong>Webhook:</strong> <code>POST {self.webhook_path}</code><br>
                    <strong>Health Check:</strong> <code>GET /health</code>
                </div>

                <h2>Configuration:</h2>
                <ul>
                    <li><strong>Port:</strong> {self.port}</li>
                    <li><strong>Secret:</strong> {self.webhook_secret[:8]}... (truncated)</li>
                    <li><strong>Webhook ID:</strong> {self.webhook_id or 'Not created yet'}</li>
                </ul>

                <h2>Supported Events:</h2>
                <ul>
                    <li>track.uploaded</li>
                    <li>track.processed</li>
                    <li>analysis.completed</li>
                    <li>transcription.completed</li>
                    <li>distribution.live</li>
                    <li>copilot.generation_completed</li>
                </ul>
            </body>
            </html>
            """
        
        @self.app.route('/health', methods=['GET'])
        def health_check():
            """Health check endpoint."""
            return jsonify({
                'status': 'healthy',
                'timestamp': datetime.utcnow().isoformat(),
                'service': 'jewelmusic-webhook-server',
                'version': '1.0.0'
            })
        
        @self.app.route(self.webhook_path, methods=['POST'])
        def handle_webhook():
            """Handle incoming webhooks."""
            try:
                # Get signature from headers
                signature = request.headers.get('X-JewelMusic-Signature')
                if not signature:
                    print("❌ Missing webhook signature")
                    return jsonify({'error': 'Missing signature'}), 401
                
                # Get raw body
                payload = request.get_data()
                
                # Verify signature
                if not WebhookResource.verify_signature(payload, signature, self.webhook_secret):
                    print("❌ Invalid webhook signature")
                    return jsonify({'error': 'Invalid signature'}), 401
                
                # Parse event
                event = WebhookResource.parse_event(payload)
                
                # Log event
                timestamp = datetime.now().strftime("%H:%M:%S")
                print(f"\n[{timestamp}] 📨 Received webhook event: {event['type']}")
                print(f"         Event ID: {event['id']}")
                print(f"         Timestamp: {event['timestamp']}")
                
                # Process event
                self._process_webhook_event(event)
                
                # Respond with success
                return jsonify({
                    'received': True,
                    'event_type': event['type'],
                    'event_id': event['id']
                })
                
            except Exception as e:
                print(f"❌ Error processing webhook: {e}")
                return jsonify({'error': 'Bad Request'}), 400
    
    def _process_webhook_event(self, event: Dict[str, Any]):
        """Process different webhook event types."""
        try:
            event_type = event['type']
            event_data = event['data']
            
            if event_type == 'track.uploaded':
                self._handle_track_uploaded(event_data)
            elif event_type == 'track.processed':
                self._handle_track_processed(event_data)
            elif event_type == 'analysis.completed':
                self._handle_analysis_completed(event_data)
            elif event_type == 'transcription.completed':
                self._handle_transcription_completed(event_data)
            elif event_type == 'distribution.live':
                self._handle_distribution_live(event_data)
            elif event_type == 'copilot.generation_completed':
                self._handle_copilot_generation_completed(event_data)
            else:
                print(f"         ⚠️  Unknown event type: {event_type}")
                
        except Exception as e:
            print(f"❌ Error handling webhook event: {e}")
    
    def _handle_track_uploaded(self, data: Dict[str, Any]):
        """Handle track uploaded event."""
        track = data['track']
        print(f"         🎵 Track uploaded: '{track['title']}' by {track['artist']}")
        print(f"         📁 Track ID: {track['id']}")
        print(f"         ⏱️  Duration: {track['duration']}s")
    
    def _handle_track_processed(self, data: Dict[str, Any]):
        """Handle track processed event."""
        track = data['track']
        print(f"         ✅ Track processed: '{track['title']}'")
        print(f"         📁 Track ID: {track['id']}")
        print(f"         📊 Status: {track['status']}")
    
    def _handle_analysis_completed(self, data: Dict[str, Any]):
        """Handle analysis completed event."""
        analysis = data['analysis']
        print(f"         🔍 Analysis completed: {analysis['id']}")
        
        if 'track_id' in analysis:
            print(f"         🎵 Track ID: {analysis['track_id']}")
        
        if 'tempo' in analysis and 'bpm' in analysis['tempo']:
            print(f"         🥁 Tempo: {analysis['tempo']['bpm']} BPM")
        
        if 'key' in analysis:
            key_data = analysis['key']
            if 'key' in key_data and 'mode' in key_data:
                print(f"         🎹 Key: {key_data['key']} {key_data['mode']}")
    
    def _handle_transcription_completed(self, data: Dict[str, Any]):
        """Handle transcription completed event."""
        transcription = data['transcription']
        print(f"         📝 Transcription completed: {transcription['id']}")
        
        if 'track_id' in transcription:
            print(f"         🎵 Track ID: {transcription['track_id']}")
        
        if 'language' in transcription:
            print(f"         🌐 Language: {transcription['language']}")
        
        if 'text' in transcription:
            preview = transcription['text'][:100]
            if len(transcription['text']) > 100:
                preview += "..."
            print(f"         📄 Text preview: {preview}")
    
    def _handle_distribution_live(self, data: Dict[str, Any]):
        """Handle distribution live event."""
        release = data['release']
        print(f"         🚀 Release is now live: '{release['title']}'")
        print(f"         📁 Release ID: {release['id']}")
        print(f"         🎤 Artist: {release['artist']}")
        
        if 'platforms' in release:
            platforms = ', '.join(release['platforms'])
            print(f"         📱 Platforms: {platforms}")
    
    def _handle_copilot_generation_completed(self, data: Dict[str, Any]):
        """Handle copilot generation completed event."""
        generation = data['generation']
        print(f"         🤖 AI generation completed: {generation['id']}")
        
        if 'type' in generation:
            print(f"         🎯 Type: {generation['type']}")
        
        if 'title' in generation:
            print(f"         📝 Title: {generation['title']}")
        
        if 'style' in generation:
            print(f"         🎨 Style: {generation['style']}")
    
    async def setup_webhook(self) -> Dict[str, Any]:
        """Setup webhook endpoint in JewelMusic."""
        print("🔧 Setting up webhook endpoint...")
        
        try:
            webhook_url = f"http://localhost:{self.port}{self.webhook_path}"
            
            webhook = await self.client.webhooks.create(
                url=webhook_url,
                events=[
                    'track.uploaded',
                    'track.processed',
                    'analysis.completed',
                    'transcription.completed',
                    'distribution.live',
                    'copilot.generation_completed'
                ],
                secret=self.webhook_secret
            )
            
            self.webhook_id = webhook['id']
            
            print("✅ Webhook created successfully!")
            print(f"Webhook ID: {webhook['id']}")
            print(f"URL: {webhook['url']}")
            print(f"Events: {', '.join(webhook['events'])}")
            print(f"Secret: {self.webhook_secret[:8]}...")
            
            return webhook
            
        except Exception as e:
            print(f"⚠️  Failed to setup webhook: {e}")
            print("Continuing with server anyway for demonstration...")
            return {}
    
    def run(self):
        """Run the webhook server."""
        print(f"🎵 JewelMusic Webhook Server")
        print(f"===========================")
        print(f"🔑 Using API key: {self.api_key[:12]}...")
        print(f"🌐 Starting webhook server on port {self.port}")
        
        # Setup webhook endpoint
        try:
            loop = asyncio.new_event_loop()
            asyncio.set_event_loop(loop)
            webhook = loop.run_until_complete(self.setup_webhook())
        except Exception as e:
            print(f"Failed to setup webhook: {e}")
        
        print(f"\n🚀 Webhook server listening on http://localhost:{self.port}")
        print(f"📍 Webhook endpoint: http://localhost:{self.port}{self.webhook_path}")
        print(f"🏥 Health check: http://localhost:{self.port}/health")
        print("\nPress Ctrl+C to stop the server")
        
        # Handle graceful shutdown
        def signal_handler(sig, frame):
            print("\n🛑 Shutting down webhook server...")
            sys.exit(0)
        
        signal.signal(signal.SIGINT, signal_handler)
        
        # Run Flask app
        self.app.run(
            host='0.0.0.0',
            port=self.port,
            debug=False,
            use_reloader=False
        )


async def manage_webhook_deliveries():
    """Webhook delivery management example."""
    print("📦 Managing webhook deliveries...")
    
    api_key = os.getenv('JEWELMUSIC_API_KEY')
    if not api_key:
        raise ValueError("API key required.")
    
    client = JewelMusic(api_key=api_key, environment='sandbox')
    
    try:
        # Get all webhooks
        webhooks = await client.webhooks.list()
        
        if not webhooks['items']:
            print("No webhooks found")
            return
        
        webhook = webhooks['items'][0]  # Use first webhook
        
        # Get delivery history
        print("📋 Getting delivery history...")
        deliveries = await client.webhooks.get_deliveries(
            webhook['id'],
            page=1,
            per_page=10,
            sort_by='created_at',
            sort_order='desc'
        )
        
        print(f"✅ Found {deliveries['pagination']['total']} deliveries")
        
        # Show recent deliveries
        for delivery in deliveries['items'][:5]:
            print(f"- {delivery['event_type']}: {delivery['status']} ({delivery['response_status']})")
            print(f"  Attempts: {delivery['attempts']}, Last attempt: {delivery['last_attempt_at']}")
        
        # Get failed deliveries
        failed_deliveries = await client.webhooks.get_deliveries(
            webhook['id'],
            status='failed',
            per_page=5
        )
        
        if failed_deliveries['items']:
            print(f"\n❌ Failed deliveries found: {len(failed_deliveries['items'])}")
            
            # Retry failed deliveries
            for delivery in failed_deliveries['items']:
                print(f"🔄 Retrying delivery: {delivery['id']}")
                await client.webhooks.retry_delivery(webhook['id'], delivery['id'])
        
        # Get delivery details
        if deliveries['items']:
            latest_delivery = deliveries['items'][0]
            print("\n📝 Latest delivery details:")
            
            delivery_details = await client.webhooks.get_delivery(
                webhook['id'],
                latest_delivery['id']
            )
            
            print(f"Event type: {delivery_details['event_type']}")
            print(f"Status: {delivery_details['status']}")
            print(f"Response status: {delivery_details['response']['status']}")
            print(f"Response time: {delivery_details['response_time']}ms")
            
    except Exception as e:
        print(f"❌ Error managing webhook deliveries: {e}")


def create_test_signature():
    """Create test signature for development."""
    print("🧪 Testing signature creation and verification...")
    
    test_payload = json.dumps({
        'id': 'evt_test_123',
        'type': 'track.uploaded',
        'timestamp': datetime.utcnow().isoformat(),
        'data': {'track': {'id': 'track_123', 'title': 'Test Song'}}
    }).encode('utf-8')
    
    secret = 'test_webhook_secret'
    
    # Create signature
    signature = WebhookResource.create_signature(test_payload, secret)
    print(f"Test signature created: {signature}")
    
    # Verify signature
    is_valid = WebhookResource.verify_signature(test_payload, signature, secret, tolerance=300)
    print(f"Signature validation: {'✅ Valid' if is_valid else '❌ Invalid'}")
    
    return signature


async def webhook_management_example():
    """Comprehensive webhook management example."""
    print("🔗 JewelMusic Webhook Management Examples")
    print("=========================================")
    
    api_key = os.getenv('JEWELMUSIC_API_KEY')
    if not api_key:
        raise ValueError("API key required.")
    
    client = JewelMusic(api_key=api_key, environment='sandbox')
    
    try:
        # List existing webhooks
        print("📋 Listing existing webhooks...")
        webhooks = await client.webhooks.list()
        print(f"Found {len(webhooks['items'])} existing webhooks:")
        
        for webhook in webhooks['items']:
            print(f"- {webhook['id']}: {webhook['url']}")
            print(f"  Events: {', '.join(webhook['events'])}")
            print(f"  Active: {webhook['active']}")
        
        # Test a webhook if one exists
        if webhooks['items']:
            webhook = webhooks['items'][0]
            print(f"\n🧪 Testing webhook: {webhook['id']}")
            
            test_result = await client.webhooks.test(webhook['id'])
            print(f"✅ Webhook test result:")
            print(f"Success: {test_result['success']}")
            print(f"Response status: {test_result['response_status']}")
            print(f"Response time: {test_result['response_time']}ms")
        
        # Get webhook statistics
        if webhooks['items']:
            webhook = webhooks['items'][0]
            print(f"\n📊 Getting webhook statistics...")
            
            try:
                stats = await client.webhooks.get_statistics(
                    webhook['id'],
                    period='last_7_days',
                    group_by='day'
                )
                
                print(f"✅ Webhook statistics:")
                print(f"Total deliveries: {stats['total_deliveries']}")
                print(f"Success rate: {stats['success_rate'] * 100:.1f}%")
                print(f"Average response time: {stats['average_response_time']}ms")
                
            except Exception as e:
                print(f"⚠️  Statistics not available: {e}")
        
    except Exception as e:
        print(f"❌ Error in webhook management: {e}")


def main():
    """Main function."""
    print("🔗 JewelMusic Webhook Handling Examples")
    print("=======================================\n")
    
    # Check for API key
    if not os.getenv('JEWELMUSIC_API_KEY'):
        print("❌ JEWELMUSIC_API_KEY environment variable not set")
        print("Please set your API key: export JEWELMUSIC_API_KEY=your_key_here")
        return 1
    
    # Get port from environment
    port = int(os.getenv('PORT', 8080))
    
    try:
        # Test signature functionality
        create_test_signature()
        print()
        
        # Run webhook management examples
        asyncio.run(webhook_management_example())
        print()
        
        # Start webhook server
        print("🌐 Starting webhook server...")
        server = WebhookServer(port=port)
        
        print("✅ Webhook handling example setup complete!")
        print("🔧 To test webhooks:")
        print("1. Make sure your webhook URL is publicly accessible")
        print("2. Use ngrok for local testing: ngrok http", port)
        print("3. Update webhook URL in JewelMusic dashboard")
        print("4. Trigger events by uploading tracks or other actions")
        print()
        
        # Run the server (this will block)
        server.run()
        
        return 0
        
    except KeyboardInterrupt:
        print("\n🛑 Webhook server stopped by user")
        return 0
    except Exception as e:
        print(f"\n💥 Webhook handling example failed: {e}")
        return 1


if __name__ == '__main__':
    exit_code = main()
    sys.exit(exit_code)

File: go/examples/webhook_server.go
package main

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"os/signal"
	"strconv"
	"syscall"
	"time"

	"github.com/jewelmusic/sdk/go/jewelmusic"
)

const (
	defaultPort = 8080
	webhookPath = "/webhooks/jewelmusic"
)

func main() {
	fmt.Println("🔗 JewelMusic Go SDK - Webhook Server Example")
	fmt.Println("============================================")

	// Check for API key
	apiKey := os.Getenv("JEWELMUSIC_API_KEY")
	if apiKey == "" {
		log.Fatal("❌ JEWELMUSIC_API_KEY environment variable not set")
	}

	// Get port from environment or use default
	port := defaultPort
	if portStr := os.Getenv("PORT"); portStr != "" {
		if p, err := strconv.Atoi(portStr); err == nil {
			port = p
		}
	}

	fmt.Printf("🔑 Using API key: %s...\n", apiKey[:12])
	fmt.Printf("🌐 Starting webhook server on port %d\n", port)

	// Initialize the client
	client := jewelmusic.NewClient(apiKey)
	ctx := context.Background()

	// Setup webhook server
	server := &WebhookServer{
		client: client,
		port:   port,
		secret: generateWebhookSecret(),
	}

	// Create webhook endpoint
	if err := server.setupWebhook(ctx); err != nil {
		log.Printf("⚠️  Failed to setup webhook: %v\n", err)
		fmt.Println("Continuing with server anyway for demonstration...")
	}

	// Start the server
	if err := server.start(); err != nil {
		log.Fatalf("💥 Failed to start server: %v", err)
	}
}

type WebhookServer struct {
	client *jewelmusic.Client
	port   int
	secret string
	server *http.Server
}

func (ws *WebhookServer) setupWebhook(ctx context.Context) error {
	fmt.Println("\n🔧 Setting up webhook endpoint...")

	// In a real scenario, you'd use your actual server URL
	webhookURL := fmt.Sprintf("http://localhost:%d%s", ws.port, webhookPath)
	
	webhook, err := ws.client.Webhooks.Create(ctx, &jewelmusic.Webhook{
		URL: webhookURL,
		Events: []string{
			"track.uploaded",
			"track.processed",
			"analysis.completed",
			"transcription.completed",
			"distribution.live",
			"copilot.generation_completed",
		},
		Secret: ws.secret,
	})
	if err != nil {
		return err
	}

	fmt.Printf("✅ Webhook created successfully!\n")
	fmt.Printf("Webhook ID: %s\n", webhook.ID)
	fmt.Printf("URL: %s\n", webhook.URL)
	fmt.Printf("Events: %v\n", webhook.Events)
	fmt.Printf("Secret: %s...\n", ws.secret[:8])

	return nil
}

func (ws *WebhookServer) start() error {
	// Setup HTTP routes
	mux := http.NewServeMux()
	mux.HandleFunc("/", ws.handleRoot)
	mux.HandleFunc("/health", ws.handleHealth)
	mux.HandleFunc(webhookPath, ws.handleWebhook)

	// Create server
	ws.server = &http.Server{
		Addr:         fmt.Sprintf(":%d", ws.port),
		Handler:      mux,
		ReadTimeout:  10 * time.Second,
		WriteTimeout: 10 * time.Second,
		IdleTimeout:  60 * time.Second,
	}

	// Start server in goroutine
	go func() {
		fmt.Printf("\n🚀 Webhook server listening on http://localhost:%d\n", ws.port)
		fmt.Printf("📍 Webhook endpoint: http://localhost:%d%s\n", ws.port, webhookPath)
		fmt.Printf("🏥 Health check: http://localhost:%d/health\n", ws.port)
		fmt.Println("\nPress Ctrl+C to stop the server")

		if err := ws.server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
			log.Printf("💥 Server error: %v", err)
		}
	}()

	// Wait for interrupt signal to gracefully shutdown
	return ws.waitForShutdown()
}

func (ws *WebhookServer) waitForShutdown() error {
	quit := make(chan os.Signal, 1)
	signal.Notify(quit, syscall.SIGINT, syscall.SIGTERM)
	<-quit

	fmt.Println("\n🛑 Shutting down webhook server...")

	// Graceful shutdown with timeout
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	if err := ws.server.Shutdown(ctx); err != nil {
		return fmt.Errorf("server forced to shutdown: %w", err)
	}

	fmt.Println("✅ Webhook server stopped gracefully")
	return nil
}

func (ws *WebhookServer) handleRoot(w http.ResponseWriter, r *http.Request) {
	if r.URL.Path != "/" {
		http.NotFound(w, r)
		return
	}

	html := `
<!DOCTYPE html>
<html>
<head>
    <title>JewelMusic Webhook Server</title>
    <style>
        body { font-family: Arial, sans-serif; margin: 40px; }
        .header { color: #333; }
        .endpoint { background: #f5f5f5; padding: 10px; border-radius: 5px; }
        .status { color: #28a745; }
    </style>
</head>
<body>
    <h1 class="header">🎵 JewelMusic Webhook Server</h1>
    <p class="status">✅ Server is running and ready to receive webhooks</p>
    
    <h2>Endpoints:</h2>
    <div class="endpoint">
        <strong>Webhook:</strong> <code>POST %s</code><br>
        <strong>Health Check:</strong> <code>GET /health</code>
    </div>

    <h2>Configuration:</h2>
    <ul>
        <li><strong>Port:</strong> %d</li>
        <li><strong>Secret:</strong> %s... (truncated)</li>
    </ul>

    <h2>Supported Events:</h2>
    <ul>
        <li>track.uploaded</li>
        <li>track.processed</li>
        <li>analysis.completed</li>
        <li>transcription.completed</li>
        <li>distribution.live</li>
        <li>copilot.generation_completed</li>
    </ul>
</body>
</html>`

	w.Header().Set("Content-Type", "text/html")
	fmt.Fprintf(w, html, webhookPath, ws.port, ws.secret[:8])
}

func (ws *WebhookServer) handleHealth(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}

	response := map[string]interface{}{
		"status":    "healthy",
		"timestamp": time.Now().UTC().Format(time.RFC3339),
		"service":   "jewelmusic-webhook-server",
		"version":   "1.0.0",
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusOK)
	fmt.Fprintf(w, `{"status":"%s","timestamp":"%s","service":"%s","version":"%s"}`, 
		response["status"], response["timestamp"], response["service"], response["version"])
}

func (ws *WebhookServer) handleWebhook(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}

	// Read the request body
	body, err := io.ReadAll(r.Body)
	if err != nil {
		log.Printf("❌ Failed to read request body: %v", err)
		http.Error(w, "Failed to read request body", http.StatusBadRequest)
		return
	}
	defer r.Body.Close()

	// Get the signature from headers
	signature := r.Header.Get("X-JewelMusic-Signature")
	if signature == "" {
		log.Println("❌ Missing webhook signature")
		http.Error(w, "Missing webhook signature", http.StatusBadRequest)
		return
	}

	// Verify webhook signature
	if !jewelmusic.VerifyWebhookSignature(body, signature, ws.secret) {
		log.Println("❌ Invalid webhook signature")
		http.Error(w, "Invalid webhook signature", http.StatusUnauthorized)
		return
	}

	// Parse webhook event
	event, err := jewelmusic.ParseWebhookEvent(body)
	if err != nil {
		log.Printf("❌ Failed to parse webhook event: %v", err)
		http.Error(w, "Failed to parse webhook event", http.StatusBadRequest)
		return
	}

	// Process the webhook event
	if err := ws.processWebhookEvent(event); err != nil {
		log.Printf("❌ Failed to process webhook event: %v", err)
		http.Error(w, "Failed to process webhook event", http.StatusInternalServerError)
		return
	}

	// Respond with success
	w.WriteHeader(http.StatusOK)
	fmt.Fprintf(w, `{"received": true, "event_type": "%s", "event_id": "%s"}`, 
		event.Type, event.ID)
}

func (ws *WebhookServer) processWebhookEvent(event *jewelmusic.WebhookEvent) error {
	timestamp := time.Now().Format("15:04:05")
	
	fmt.Printf("\n[%s] 📨 Received webhook event: %s\n", timestamp, event.Type)
	fmt.Printf("         Event ID: %s\n", event.ID)
	fmt.Printf("         Timestamp: %s\n", event.Timestamp.Format(time.RFC3339))

	// Handle different event types
	switch event.Type {
	case "track.uploaded":
		return ws.handleTrackUploaded(event)
	case "track.processed":
		return ws.handleTrackProcessed(event)
	case "analysis.completed":
		return ws.handleAnalysisCompleted(event)
	case "transcription.completed":
		return ws.handleTranscriptionCompleted(event)
	case "distribution.live":
		return ws.handleDistributionLive(event)
	case "copilot.generation_completed":
		return ws.handleCopilotGenerationCompleted(event)
	default:
		fmt.Printf("         ⚠️  Unknown event type: %s\n", event.Type)
		return nil
	}
}

func (ws *WebhookServer) handleTrackUploaded(event *jewelmusic.WebhookEvent) error {
	track, ok := event.Data["track"].(map[string]interface{})
	if !ok {
		return fmt.Errorf("invalid track data in webhook")
	}

	fmt.Printf("         🎵 Track uploaded: '%s' by %s\n", 
		track["title"], track["artist"])
	fmt.Printf("         📁 Track ID: %s\n", track["id"])
	fmt.Printf("         ⏱️  Duration: %.0fs\n", track["duration"])

	return nil
}

func (ws *WebhookServer) handleTrackProcessed(event *jewelmusic.WebhookEvent) error {
	track, ok := event.Data["track"].(map[string]interface{})
	if !ok {
		return fmt.Errorf("invalid track data in webhook")
	}

	fmt.Printf("         ✅ Track processed: '%s'\n", track["title"])
	fmt.Printf("         📁 Track ID: %s\n", track["id"])
	fmt.Printf("         📊 Status: %s\n", track["status"])

	return nil
}

func (ws *WebhookServer) handleAnalysisCompleted(event *jewelmusic.WebhookEvent) error {
	analysis, ok := event.Data["analysis"].(map[string]interface{})
	if !ok {
		return fmt.Errorf("invalid analysis data in webhook")
	}

	fmt.Printf("         🔍 Analysis completed: %s\n", analysis["id"])
	
	if trackID, exists := analysis["track_id"]; exists {
		fmt.Printf("         🎵 Track ID: %s\n", trackID)
	}
	
	if tempo, exists := analysis["tempo"]; exists {
		if tempoMap, ok := tempo.(map[string]interface{}); ok {
			if bpm, exists := tempoMap["bpm"]; exists {
				fmt.Printf("         🥁 Tempo: %.1f BPM\n", bpm)
			}
		}
	}
	
	if key, exists := analysis["key"]; exists {
		if keyMap, ok := key.(map[string]interface{}); ok {
			if keyName, exists := keyMap["key"]; exists {
				if mode, exists := keyMap["mode"]; exists {
					fmt.Printf("         🎹 Key: %s %s\n", keyName, mode)
				}
			}
		}
	}

	return nil
}

func (ws *WebhookServer) handleTranscriptionCompleted(event *jewelmusic.WebhookEvent) error {
	transcription, ok := event.Data["transcription"].(map[string]interface{})
	if !ok {
		return fmt.Errorf("invalid transcription data in webhook")
	}

	fmt.Printf("         📝 Transcription completed: %s\n", transcription["id"])
	
	if trackID, exists := transcription["track_id"]; exists {
		fmt.Printf("         🎵 Track ID: %s\n", trackID)
	}
	
	if language, exists := transcription["language"]; exists {
		fmt.Printf("         🌐 Language: %s\n", language)
	}
	
	if text, exists := transcription["text"]; exists {
		if textStr, ok := text.(string); ok {
			preview := textStr
			if len(preview) > 100 {
				preview = preview[:100] + "..."
			}
			fmt.Printf("         📄 Text preview: %s\n", preview)
		}
	}

	return nil
}

func (ws *WebhookServer) handleDistributionLive(event *jewelmusic.WebhookEvent) error {
	release, ok := event.Data["release"].(map[string]interface{})
	if !ok {
		return fmt.Errorf("invalid release data in webhook")
	}

	fmt.Printf("         🚀 Release is now live: '%s'\n", release["title"])
	fmt.Printf("         📁 Release ID: %s\n", release["id"])
	fmt.Printf("         🎤 Artist: %s\n", release["artist"])
	
	if platforms, exists := release["platforms"]; exists {
		if platformList, ok := platforms.([]interface{}); ok {
			fmt.Printf("         📱 Platforms: ")
			for i, platform := range platformList {
				if i > 0 {
					fmt.Printf(", ")
				}
				fmt.Printf("%s", platform)
			}
			fmt.Println()
		}
	}

	return nil
}

func (ws *WebhookServer) handleCopilotGenerationCompleted(event *jewelmusic.WebhookEvent) error {
	generation, ok := event.Data["generation"].(map[string]interface{})
	if !ok {
		return fmt.Errorf("invalid generation data in webhook")
	}

	fmt.Printf("         🤖 AI generation completed: %s\n", generation["id"])
	
	if genType, exists := generation["type"]; exists {
		fmt.Printf("         🎯 Type: %s\n", genType)
	}
	
	if title, exists := generation["title"]; exists {
		fmt.Printf("         📝 Title: %s\n", title)
	}
	
	if style, exists := generation["style"]; exists {
		fmt.Printf("         🎨 Style: %s\n", style)
	}

	return nil
}

// generateWebhookSecret generates a secure random secret for webhook verification
func generateWebhookSecret() string {
	bytes := make([]byte, 32)
	if _, err := rand.Read(bytes); err != nil {
		// Fallback to a default secret if random generation fails
		return "default_webhook_secret_change_in_production"
	}
	return hex.EncodeToString(bytes)
}

// Example of how to use the webhook server programmatically
func exampleProgrammaticUsage() {
	client := jewelmusic.NewClient(os.Getenv("JEWELMUSIC_API_KEY"))
	ctx := context.Background()

	// List existing webhooks
	webhooks, err := client.Webhooks.List(ctx)
	if err != nil {
		log.Printf("Failed to list webhooks: %v", err)
		return
	}

	fmt.Printf("Found %d existing webhooks:\n", len(webhooks.Items))
	for _, webhook := range webhooks.Items {
		fmt.Printf("- %s: %s\n", webhook.ID, webhook.URL)
	}

	// Test a webhook
	if len(webhooks.Items) > 0 {
		webhook := webhooks.Items[0]
		err := client.Webhooks.Test(ctx, webhook.ID)
		if err != nil {
			log.Printf("Failed to test webhook: %v", err)
		} else {
			fmt.Printf("✅ Webhook test sent to %s\n", webhook.URL)
		}
	}
}